submariner-operator/issue-149-node-placement.md

```diff
@@ -0,0 +1,94 @@
+# Node Placement
+
+[Add support for nodeSelector and toleration scheduling for submariner pods](https://github.com/submariner-io/enhancements/issues/149)
+
+## Summary
+
+Submariner deploys many different agents as part of its core deployment. Gateway, Globalnet Agent and
+Metrics Proxy are deployed on designated gateway nodes, while Route Aents are deployed on all nodes. There is no way
+for users to control or limit where the rest of the agent pods are deployed. Some users want to run these agents on
+specific nodes for billing or networking reasons. Users can't add tolerations directly to these agents because they'll be
+overwritten by the operator. So, users require an easily configurable way to specify these parameters.
+
+## Proposal
+
+This enhancement proposes adding a new field to `Submariner` and `ServiceDiscovery` for users to configure node
+scheduling. This configuration will then be applied to pods of following submariner agents:
+
+* submariner-operator
+* submariner-lighthouse-agent
+* submariner-lighthouse-coredns
+* submariner-networkplugin-syncer
+
+Following agents will be excluded because these are run on Gateway nodes. Users already control which nodes get labeled
+as gateways and ideally we don't want to run anything else on these nodes.
+
+* submariner-gateway
+* submariner-globalnet
+* submariner-metrics-proxy
+
+## Design Details
+
+Add new `NodeSelector` and `Tolerations` fields to `Submariner` and `ServiceDiscovery` specs.
+
+```Go
+type SubmarinerSpec struct {
+    ...
+    ... 
+    // NodeSelector defines which Nodes the Pods are scheduled on. The default is an empty list.
+    // +optional
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+    // Tolerations is attached by pods to tolerate any taint that matches
+    // the triple <key,value,effect> using the matching operator <operator>.
+    // The default is an empty list.
+    // +optional
+    Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+}
+```
+
+```Go
+type SubmarinerDiscoverySpec struct {
+    ...
+    ...
+    // NodeSelector defines which Nodes the Pods are scheduled on. The default is an empty list.
+    // +optional
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+    // Tolerations is attached by pods to tolerate any taint that matches
+    // the triple <key,value,effect> using the matching operator <operator>.
+    // The default is an empty list.
+    // +optional
+    Tolerations []v1.Toleration `json:"tolerations,omitempty"`
+}
+```
+
+Add option to `subctl join` to specify node selectors and tolerations.
+
+```bash
+subctl join --node-selectors LABEL_KEY_1=LABEL_VALUE_1,...,LABEL_KEY_N=LABEL_VALUE_N
+subctl join --node-tolerations <path-to-yaml-file>
```

Comment:
```suggestion
subctl join --tolerations <path-to-yaml-file>
```

Shouldn't it be just `--tolerations` then?

Reply:
Also if we're already talking about YAML, perhaps we can just let users specify fields via YAML directly then?

Reply:
@mkolesnik You mean allow node selector and tolerations both through yaml?